xwiki-platform-core/xwiki-platform-javascript/xwiki-platform-javascript-importmap/src/main/java/org/xwiki/javascript/importmap/internal/JavascriptImportmapResolver.java

```diff
@@ -0,0 +1,187 @@
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.extension.Extension;
+import org.xwiki.extension.repository.CoreExtensionRepository;
+import org.xwiki.extension.repository.InstalledExtensionRepository;
+import org.xwiki.rendering.block.Block;
+import org.xwiki.rendering.block.RawBlock;
+import org.xwiki.webjars.WebJarsUrlFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+
+import static com.fasterxml.jackson.databind.MapperFeature.SORT_PROPERTIES_ALPHABETICALLY;
+import static org.apache.commons.lang3.exception.ExceptionUtils.getRootCauseMessage;
+import static org.xwiki.rendering.syntax.Syntax.HTML_5_0;
+
+/**
+ * Resolve the importmap from the webjar declarations, using the {@code "xwiki.javascript.modules.importmap"} pom
+ * property.
+ *
+ * @version $Id$
+ * @since 17.10.0RC1
+ */
+@Component(roles = JavascriptImportmapResolver.class)
+@Singleton
+public class JavascriptImportmapResolver
+{
+    private static final ObjectMapper OBJECT_MAPPER = JsonMapper.builder()
+        // Keeps the output deterministic.
+        .configure(SORT_PROPERTIES_ALPHABETICALLY, true)
+        .build();
+
+    private static final String JAVASCRIPT_IMPORTMAP_PROPERTY = "xwiki.javascript.modules.importmap";
+
+    /**
+     * Cache lock to prevent race conditions when retrieving cached values.
+     */
+    private final Object cacheLock = new Object();
+
+    /**
+     * The cached value, concurrent operation on this field must use the {@link #cacheLock} object for synchronization.
+     */
+    private Block cachedValue;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private InstalledExtensionRepository installedExtensionRepository;
+
+    @Inject
+    private CoreExtensionRepository coreExtensionRepository;
+
+    @Inject
+    private WebJarsUrlFactory webJarsUrlFactory;
+
+    /**
+     * @return the block to render from the importmap, possibly cached
+     */
+    public Block getBlock()
+    {
+        synchronized (this.cacheLock) {
+            if (this.cachedValue == null) {
+                compute();
+            }
+            return this.cachedValue;
+        }
+    }
+
+    /**
+     * Clear the cache of the resolver.
+     */
+    public void clearCache()
+    {
+        synchronized (this.cacheLock) {
+            this.cachedValue = null;
+        }
+    }
+
+    private void compute()
+    {
+        List<Map<String, String>> extensionsWithImportMap = Stream.concat(
+                this.installedExtensionRepository.getInstalledExtensions().stream(),
+                this.coreExtensionRepository.getCoreExtensions().stream())
+            .filter(extension -> accessProperty(extension) != null)
+            .map(extension -> {
+                String importMapJSON = accessProperty(extension);
```

Comment:
Also note that a good part of this logic will be moved to commons since I also plan to add an enforcer in `xwiki-commons-tool-enforcer-dependencies`
I'll likely also have to move XWIKI-23609 to XCOMMONS on Jira to account for the introduction of code in xwiki-commons.